### ISSUE TITLE
over the counter channel contracts

### ISSUE THREAD

**Issue by zack-bitcoin** [zack-bitcoin/light-node-amoveo#4]
People should be able to use a light node to make channels with custom bets directly between each other.

Benefits of this kind of betting vs the markets we already have:
* If people make bets directly with each other, they have more privacy.
* if we don't need a central server to operate betting, then there is less legal cost. Because there is no single point of weakness that can be legally exploited.
* if people make bets in direct 1-step channel paths, then there is less capital locked up to enforce the contract. With markets, the server needs to lock up a lot of capital to enforce the contracts.

Costs of this kind of betting vs the markets we already have:
* Over the counter type trading doesn't have as clear of a market price, so there is more loss due to arbitrage.

When does it make sense to use over-the-counter style betting?
* If the market wont resolve for a very long period of time, then OTC channel bets would be cheaper, because you don't have to pay a server to keep money locked up.
* If there are very few people who want to trade in this contract, then the market has less of an advantage at finding the market price, so in this situation OTC will often be cheaper. 
* If there is a lot of legal risk in operating a market, then it might be cheaper to avoid those legal costs and use OTC channel betting.

Todo:
* tool for generating transactions to:
   - open a new channel
   - team-close a channel
   - solo-close a channel using a valid channel state.
   - slashing a channel using a valid channel state.
   - channel timeout tx
* tool for interpreting a channel state in a human readable way.
   - it already exists, but we might need to add security features to prevent smart contract injection attacks.
   - in bets.js outstanding_bets3, we should probably try re-compiling each contract to verify that the bytes are what we expect. We should display a warning message if there is any smart contract we don't know how to handle.
* channel state editor.
  - adding bets. use market.js to generate the contract.
  - removing bets.
  - changing who owns how much veo
* sanity tool to do some basic checks that this is a valid channel state update.
   - maybe it should display a diff between 2 channel states, so you can easily see what has changed.

**Comment by zack-bitcoin** [comment]
The design outlined in the previous comment is very powerful. Users would have a lot of control over their channels that way.
But I am worried that the user experience will be too difficult.
There are just too many steps.

So here is an alternative strategy:

We could have some servers who's only job is to send messages between different light nodes.
So each participant in a otc bet will fill out a form with identical information, and then click a "start" button.
Both light nodes will keep sending messages back and forth until the contract is live. The light nodes could display a message of what step they are on, and approximately how much time is left.

**Comment by zack-bitcoin** [comment]
Ive been thinking about how to resolve the channels when a bet gets settled. It isn't reasonable to expect both participants to come online at the same time.

Maybe they should email the final tx to each other for the first version?

**Comment by zack-bitcoin** [comment]
Information on the form:
both participant's pubkeys.
you need to load your private key into the light node.
the oracle being bet on.
The quantity of veo each participant is going to bet.


Ideally it should be possible for one person to fill out the form, and then send a file to the other person so that their form gets auto-filled with the same data, and all they have to do is confirm that it is correct.
So one person fills out a long form, and the other person only needs to put the 2 pubkeys, and the rest auto-fills.

**Comment by zack-bitcoin** [comment]
Timeline

1) both fill out identical forms
2) channel state is signed by both
->
<-
3) both sign a tx to make the channel
->
4) the oracle gets closed

**Comment by zack-bitcoin** [comment]
For the first 3 messages, I am thinking we will use a messaging protocol like this:
The server charges a flat fee, and stores the messages for up to 10 minutes. All messages are encrypted.

For creating the channel team close, it might be the case that one of the users is off-line. So we need to keep the message around for possibly days.
To keep costs low, this message will not be encrypted. This isn't a privacy vulnerability, the tx is going on-chain anyway. This way the server can limit itself to holding 1 of this type of message for every account in a channel. So there is a limit of how many of this message could exist at once, so it isn't possible to overflow the server's memory with this kind of message.

**Comment by zack-bitcoin** [comment]
Now I am thinking about the time-line experience from each participant's perspective.
So in our example, Alice is running a server that Bob and Carol are using to send messages to each other to make a bet in a channel.

### Alice
She has 2 gen_servers, encrypted_mail, and channel_close_messages.
She has 4 api for this, 1 to put a message in encrypted_mail, which costs a fee, 1 to read a message from encrypted_mail. Even after being read, the message is stored until the 10 minutes are up.
1 api to send a channel_close_message. The tx needs to be signed, we only store one of these per account in each channel, so at most, we could store (number of channels)*2 of these at a time. They are deleted after a couple weeks.
1 api to read a channel_close_message, anyone can read it, we don't delete it, even if it has been read. If the channel gets closed, we can delete this message early.

### Bob
1) Bob opens his light node and loads his private key.
2) Bob fills out the form to make a trade, and he saves a CSV version to file.
3) Bob sends the CSV file to Carol by email or a forum.
4) The light node generates the smart contract, signs it, and sends the signature to Carol as an encrypted_message in Alice.
5) The light node checks every 10 seconds if Alice has  the signature sent to him from Carol.
6) The light node makes a big warning, telling Bob that he needs to save the signed smart contract to a file.
7) The light node signs a tx to make the channel, send it as an encrypted message to Carol.
8) The light node checks every 10 seconds until it sees that the new tx has been included in a block.
9) The light node makes a message telling Bob that the contract is now active, that it is now safe to shut down the lightnode, that he can delete the CSV file, and reminds him that he needs to save the smart contract from step (6).
10) (same as Carol's 10)
11) (same as Carol's 11)
12) (same as Carol's 12)
13) (same as Carol's 13)

### Carol
1) Carol receives the CSV file from Bob. 
2) She opens it in her light node so that the same information as Bob can auto-fill the form. 
3) She verifies that the information is correct and clicks the "this info is correct" button within 10 minutes of the moment when Bob clicks "start".
4) (same as Bob's step 4)
5) (same as Bob's step 5)
6) (same as Bob's step 6)
7) the lightnode signs the tx and publishes it to the blockchain.
8) the lightnode makes a message saying that it is now safe to shut off, and that the CSV file can be safely deleted. But it is important to keep the file from step (6) until the contract is completed.
9) Carol heard that the oracle has been settled, so she turns her light node on to close the channel. She loads the file from step (6). The light node automatically notices that this contract is ready to be closed.
10) the light node checks if the channel is already closed. if it is, { it tells Alice about the final state of the channel, and tells her that it is now safe to delete the file from step (6). } otherwise go to step 11.
11) The light node checks the channel_close_messages on Alice's server. If Bob sent a signed channel_close_tx, and it has the correct final state, then { Carol's light node signs and publishes the tx, it says a message about how the channel is done, it is safe to delete the file from step (6) } else go to step (12).
12)  Bob hasn't sent the message to close the channel, so Carol's light node sends the signed tx to Bob as a channel_close_message in Alice's server. 
13) Carol's light node gives Carol a message saying that we are waiting on Bob to close the channel. It gives her a button to start closing the channel without Bob's help, along with a warning about how it will break privacy and cost a larger fee, and it will probably take longer than just waiting for Bob to sign the tx. The light node tells her to keep a copy of the file from step (6) until the channel is closed.

**Comment by zack-bitcoin** [comment]
I am thinking sending the CSV file around is too messy, and that it could break privacy.
Maybe we should use an encrypted_message on Alice for this.

Lets say Bob sends an offer of a trade to Carol's address, then Alice will deliver the encrypted trade request to Carol, and Carol's light node should display a list of offers that Carol can click on and read details about.

So Carol would need to send her address to Bob using email or a forum or something, and the rest of the communication can happen from inside the light node.

**Comment by zack-bitcoin** [comment]
https://github.com/zack-bitcoin/amoveo-feed/blob/master/apps/amoveo-feed/src/accounts.erl

I think the messaging system should be a different github repository, a different erlang program that you can run at the same time as amoveo.
Similar to the amoveo-feed program, where users can pay veo to make posts on a public feed.
We can probably reuse a lot of code from amoveo-feed.

**Comment by zack-bitcoin** [comment]
https://github.com/zack-bitcoin/amoveo-messenger
I made the new repository for the messaging system. It starts as a clone of amoveo-feed.

**Comment by zack-bitcoin** [comment]
Amoveo messenger system is good enough that we can get started on the light node javascript parts.

**Comment by zack-bitcoin** [comment]
The light-node can now create a new channel with a custom bet directly between pairs of users.
We still do not have tools for closing this channels and resolving the bet.

**Comment by zack-bitcoin** [comment]
I got the binary derivatives working.
The next step is for some of us to test this out in the testnet, to make sure it acts how we expect it to in all the edge cases. https://github.com/zack-bitcoin/light-node-amoveo

Ill add some documentation now.
I added some documentation to the readme.
I installed the message passing software on the server, and set it up in testnet mode.
So now we should be able to test P2P derivatives in the testnet.
For now it only supports binary derivatives, but I will add scalar soon.
This is an exciting day. I hope that these P2P derivatives are the feature that achieves product market fit for Amoveo.
